src/metal/scheduler.ts

```diff
@@ -28,26 +28,45 @@ const TOKEN_SEED = 'xxxx'.replace(/[xy]/g, function(c) {
   return v.toString(16);
 });
 let tokenCounter = 0;
-
-function generateRandomToken() {
-  return Math.floor(Math.random() * (9999999 - 0o0)).toString(16);
-}
+let elementSchedulerCounter = 0;
 
 export class Frame implements FrameInterface {
   constructor(
     public timestamp: number,
     public scrollTop: number,
     public scrollLeft: number,
     public width: number,
-    public height: number
+    public height: number,
+    public x: number,
+    public y: number,
+    public top: number,
+    public left: number
   ) {}
-  static generate(): Frame {
+  static generate(root?: Element): Frame {
+    if (root && document.documentElement.contains(root)) {
```

Comment:
In the past, we have used the window proxy for any direct touching of the document or the window, so that things don’t break in node land. Is this a safe direct document reference?

Reply:
Nice catch. Besides the private conditionals within the window proxy for `hasDomSetup` and `hasDOM` we aren't currently (to my knowledge) checking for DOM before leveraging the `document`. We should definitely think about exporting a document proxy or even just a document interface to check for DOM etc.